docs: adds some guide on the different implementation approaches (#693)
SAP/crossplane-provider-btp: docs/contribution-notes/resource-implementation-approaches.md

# Resource Implementation Approaches

> We generally distinguish between 3 different approaches when imeplementing controller logic for a single resource. 

Reviewer: ```suggestion
We generally distinguish between 3 different approaches when implementing controller logic for a single resource. 
```

Next to, what we call "native controllers", you will find another approach based on upjet as well as a hybrid one throughout the provider implementation.
The choice depends on whether a native BTP API exists for the resource and how much custom logic is needed around lifecycle management.

> Lets have a closer look at those different approaches.

Reviewer 2: ```suggestion
Let's have a closer look at those different approaches.
```

> **Diagram color coding:** Throughout this document, component colors indicate their origin: **red** = crossplane-runtime, **white** = provider implementation, **blue** = upjet.

## 1. Native Resources
The controller is implemented directly in Go and talks to BTP APIs 

### How it works
![Native controller concept](../img/concept-native-controller.png)

The controller is hand-written and owns the full reconciliation logic — it decides what "up to date" means, how to handle drift, and how to map CR fields to API calls.

> The API calls themselfs are usually done via a generated api client of some sort.

Reviewer 2: ```suggestion
The API calls themselves are usually done via a generated api client of some sort.
```

> ### Resources using this approach

Reviewer 2: Should we mention `GlobalAccount` and `ResourceUsage` here? Or are these special cases?

Author: Good question. I don't think we need to care about `GlobalAccount` since its about to be removed and would be another place to remember to remove. 
`ResourceUsage` is special in the sense that its not a cloud resource. However it is a native implementation by that standards, so I will add it.

Author: Done

|---|---|
| Directory Entitlement | `btp_directory_entitlement` |
| Global Account Trust Configuration | `btp_globalaccount_trust_configuration` |
| Subaccount API Credential | `btp_subaccount_api_credential` |
| Subaccount Trust Configuration | `btp_subaccount_trust_configuration` |
| Subaccount Service Broker | `btp_subaccount_service_broker` |

---

## 3. Hybrid Resources

The user interacts with a native (meaning developer defined) CR. Internally, the controller maps that CR to an equivalent Terraform-backed type and delegates all infrastructure operations to an embedded upjet controller. The internal Terraform type is never exposed to the user.

This approach is used when a Terraform resource already exists and provides solid infrastructure management, but the user-facing API needs some adaptations or custom behaviour. Such custom behaviour might e.g. be more complex reference resolution, dependency checks or secret handling.

### How it works
![Hybrid controller concept](../img/concept-hybrid-controller.png)

> **Note:** In that diagram the blue components are autogenerated via upjet. This case shows the implementation of the ServiceInstance.

> The user-facing CRD and its controller are natively implemented. In order to handle all the API-specific aspects it wraps an upjetted controller and delegates the usual CRUD methods (Observe, Create, Update, Delete) to it. In order to do so it defines a mapping between the two CRDs. Results are translated back and written into the native CR's status.

Reviewer: I find the term `CRD` a bit off. Wouldn't MR (Managed Resource) be a better wording here?_

Author: Yes I guess you are right. MRs for me is always bound to the particular instance, but thats not actually what it is. Replacing with `MR`.

> > **Note:** The real implementation is more complex then shown here: whats referred to controller here is actually split into a connector (responsible for setup and mapping) and a controller (responsible for the CRUD calls). Some resources might also use more than one embedded upjet controller internally. There may also be additional initialization steps — for example, resolving a service plan ID before the first reconciliation.

Reviewer 2: ```suggestion
> **Note:** The real implementation is more complex than shown here: what is referred to controller here is actually split into a connector (responsible for setup and mapping) and a controller (responsible for the CRUD calls). Some resources might also use more than one embedded upjet controller internally. There may also be additional initialization steps — for example, resolving a service plan ID before the first reconciliation.
```

### Key parts of the implementation

The following sections show some of the key aspects of the implementation by using the `ServiceInstance` as the reference example.

#### Setting up the internal Terraform controller

Before any reconciliation can happen, the native connector must construct the internal upjet controller it will delegate to. This involves creating the upjet-compatible connector (backed by a Terraform workspace) and wrapping it together with the mapper into a `TfProxyConnector`.

> For `ServiceInstance` this wiring happens in [`newClientCreatorFn`](../../internal/controller/account/serviceinstance/serviceinstance.go#L53) and [`NewServiceInstanceConnector`](../../internal/clients/account/serviceinstance/serviceinstance.go#L22).

Reviewer 2: If we want to link the code directly I think we should rather point to a specific commit as the current main will diverge and the links will be broken. Alternatively we just link the file + method which is less likely to get outdated.

Author: Done
